src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -0,0 +1,312 @@
+    for (auto it = pending_tokens.begin(); it != pending_tokens.end(); it++)
+    {
+        if (it->token == token)
+        {
+            pending_tokens.erase(it);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void mf::XdgActivationV1::bind(struct wl_resource* resource)
+{
+    new Instance{resource, this, shell, focus_controller, main_loop};
+}
+
+mf::XdgActivationV1::Instance::Instance(
+    struct wl_resource* resource,
+    mf::XdgActivationV1* xdg_activation_v1,
+    std::shared_ptr<msh::Shell> const& shell,
+    std::shared_ptr<msh::FocusController> const& focus_controller,
+    std::shared_ptr<MainLoop> const& main_loop)
+    : XdgActivationV1(resource, Version<1>()),
+      xdg_activation_v1{xdg_activation_v1},
+      shell{shell},
+      focus_controller{focus_controller},
+      main_loop{main_loop}
+{
+}
+
+void mf::XdgActivationV1::Instance::get_activation_token(struct wl_resource* id)
+{
+    new XdgActivationTokenV1(id, xdg_activation_v1->create_token(), focus_controller);
+}
+
+void mf::XdgActivationV1::Instance::activate(std::string const& token, struct wl_resource* surface)
+{
+    if (!xdg_activation_v1->try_consume_token(token))
+    {
+        mir::log_error("XdgActivationV1::activate invalid token: %s", token.c_str());
+        return;
+    }
+
+    main_loop->enqueue(this, [surface=surface, shell=shell]
+    {
+        auto const wl_surface = mf::WlSurface::from(surface);
+        if (!wl_surface)
+        {
+            mir::log_error("XdgActivationV1::activate wl_surface not found");
+            return;
+        }
+
+        auto const scene_surface_opt = wl_surface->scene_surface();
+        if (!scene_surface_opt)
+        {
+            mir::log_error("XdgActivationV1::activate scene_surface_opt not found");
+            return;
+        }
+
+        auto const& scene_surface = scene_surface_opt.value();
+        auto const now = std::chrono::steady_clock::now().time_since_epoch();
+        auto ns = std::chrono::duration_cast<std::chrono::nanoseconds >(now).count();
+        shell->raise_surface(wl_surface->session, scene_surface, ns);
+    });
+}
+
+mf::XdgActivationTokenV1::XdgActivationTokenV1(
+    struct wl_resource* resource,
+    std::string const& token,
+    std::shared_ptr<msh::FocusController> const& focus_controller)
+    : wayland::XdgActivationTokenV1(resource, Version<1>()),
+      token{token},
+      focus_controller{focus_controller}
+{
+}
+
+void mf::XdgActivationTokenV1::set_serial(uint32_t serial_, struct wl_resource* seat_)
+{
+    serial = serial_;
+    seat = seat_;
+}
+
+void mf::XdgActivationTokenV1::set_app_id(std::string const& app_id_)
+{
+    app_id = app_id_;
+}
+
+void mf::XdgActivationTokenV1::set_surface(struct wl_resource* surface)
+{
+    requesting_surface = surface;
+}
+
+void mf::XdgActivationTokenV1::commit()
+{
+    // In the event of a failure, we send 'done' with a new, invalid token
+    // which cannot be used later.
```

Comment:
That doesn't accord with my reading of the protocol:

```xml
        <event name="done">
            <description summary="the exported activation token">
                The 'done' event contains the unique token of this activation request
                and notifies that the provider is done.
            </description>
            <arg name="token" type="string" summary="the exported activation token"/>
        </event>
```
Is there prior art for sending invalid tokens?

Reply:
I adapted this from wlroots, who may or may not be doing a good thing. https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/wlr_xdg_activation_v1.c?ref_type=heads#L91 The error case always sends an invalid token string

Reply:
I guess it is a case of the protocol not having "nack" event, so there's no obvious way to "reject" the request. (Making token nullable would be enough)